### PR TITLE
Amazon Q Transform: add java and maven versions to metric

### PR DESCRIPTION
**Pull request by tincheng**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

Capture the java and maven versions from the project environment in code transform metric.

example metric for project using bundled maven:
```
...metadata={codeTransformLocalMavenVersion=null (mvnw) -- 3.9.5  (mvn) -- user setting: Bundled (Maven 3), codeTransformLocalJavaVersion=JDK_1_8, ...
```

example metric for project using maven wrapper:
```
...metadata={codeTransformLocalMavenVersion=3.6.0  (mvnw) -- 3.9.5  (mvn) -- user setting: Use Maven wrapper, codeTransformLocalJavaVersion=JDK_1_8, ...
```

example metric for project using override local maven:
```
...metadata={codeTransformLocalMavenVersion=null (mvnw) -- 3.9.5  (mvn) -- user setting: 3.9.6, codeTransformLocalJavaVersion=JDK_1_8, ...
```

## Checklist
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by dhasani23 — APPROVED** [review]

**Review by JuliaABurch — APPROVED** [review]

**Comment by tincheng** [comment]
currently troubleshooting an issue where an additional IDE error notification shows when we cannot fetch maven wrapper version. We don't want to show this notification.

**Review by damntrecky — APPROVED** [review]

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 1]
```diff
@@ -616,12 +626,60 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
...
-                    codeTransformRunTimeLatency = calculateTotalLatency(CodeTransformTelemetryState.instance.getStartTime(), Instant.now())
+                    codeTransformRunTimeLatency = calculateTotalLatency(CodeTransformTelemetryState.instance.getStartTime(), Instant.now()),
+                    codeTransformLocalJavaVersion = getJavaVersionFromProjectSetting(),
+                    codeTransformLocalMavenVersion = getMavenVersions(),
                 )
             }
         }
     }
 
+    private fun getJavaVersionFromProjectSetting(): String? = project.tryGetJdk()?.toString()
```
Comment (damntrecky) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:637]: Question: Don't we already have a reference to function getting this java version from somewhere?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 2]
```diff
@@ -616,12 +626,60 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
...
+        val mavenSettings = MavenProjectsManager.getInstance(project).getGeneralSettings()
+        val mavenHome = mavenSettings.getMavenHome()
+        // TODO: Need to detect bundled Maven version that come with IDEA
+        //  The utility returns "Use Maven wrapper" if using wrapper, "Bundled (Maven 3)" if using Bundled Maven, otherwise the local maven version.
+        val userMavenSetting = MavenUtil.getMavenVersion(mavenHome) ?: mavenHome
+
+        return "$wrapperMavenVersion (mvnw) -- $localMavenVersion (mvn) -- user setting: $userMavenSetting"
+    }
+
+    private fun parseMavenVersion(output: String?): String? {
```
Comment (damntrecky) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:676]: Question: Should this function be in a utility class? The manager file shouldn't be bundled up with utility functions.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 3]
```diff
@@ -616,12 +626,60 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
...
+                    codeTransformLocalJavaVersion = getJavaVersionFromProjectSetting(),
+                    codeTransformLocalMavenVersion = getMavenVersions(),
                 )
             }
         }
     }
 
+    private fun getJavaVersionFromProjectSetting(): String? = project.tryGetJdk()?.toString()
+
+    private fun getMavenVersions(): String {
```
Comment (damntrecky) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:639]: Question: Should this function be in a utility class? The manager file shouldn't be bundled up with utility functions.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 1 above)
Comment (damntrecky) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:637]: I think `Utils.kt` in the codemodernizer folder would be a good place.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 3 above)
Comment (damntrecky) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:639]: I think `Utils.kt` in the codemodernizer folder would be a good place.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 1 above)
Comment (tincheng) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:637]: Good call, can move it to util.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 2 above)
Comment (tincheng) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:676]: Good call, can move it to util.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 3 above)
Comment (tincheng) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:639]: Good call, can move it to util.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 4]
```diff
@@ -616,12 +626,60 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
...
+        val userMavenSetting = MavenUtil.getMavenVersion(mavenHome) ?: mavenHome
+
+        return "$wrapperMavenVersion (mvnw) -- $localMavenVersion (mvn) -- user setting: $userMavenSetting"
+    }
+
+    private fun parseMavenVersion(output: String?): String? {
+        if (output == null) return null
+        val mavenVersionIndex = output.indexOf("Apache Maven")
+        val mavenVersionString = output.slice(IntRange(mavenVersionIndex + 13, output.length - 1))
+        return mavenVersionString.slice(IntRange(0, output.indexOf(' ') - 1))
```
[the remark this answers is not in the document]
Comment (tincheng) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:680]: Will put a quick fix

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt [hunk 5]
```diff
@@ -285,3 +293,50 @@ fun isCodeModernizerAvailable(project: Project): Boolean {
...
+
+    return "$wrapperMavenVersion (mvnw) -- $localMavenVersion (mvn) -- user setting: $userMavenSetting"
+}
+
+private fun parseMavenVersion(output: String?): String? {
+    if (output == null) return null
+    val mavenVersionIndex = output.indexOf("Apache Maven")
+    if (mavenVersionIndex == -1) return null
+    val mavenVersionString = output.slice(IntRange(mavenVersionIndex + 13, output.length - 1))
+    return mavenVersionString.slice(IntRange(0, output.indexOf(' ') - 1))
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt:341]:
```suggestion
    return try {
        val mavenVersionString = output.slice(IntRange(mavenVersionIndex + 13, output.length - 1))
    
        mavenVersionString.slice(IntRange(0, output.indexOf(' ') - 1))
    } catch (e: StringIndexOutOfBoundsException) {
        // also probably log an error here
        null
    }
```

Comment (tincheng) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt:341]: I think to add a Logger, I will need to introduce an object to wrap around all functions in the util file, then have to update imports across the code base, which will be a large change that doesn't seem suitable to add this PR, so I will leave the logging out.

**Comment by tincheng** [comment]
I don't want to introduce too much refactoring for all other utilities function in Util.kt, so only wrapping the three version related function in an object for now.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt [hunk 6]
```diff
@@ -285,3 +294,59 @@ fun isCodeModernizerAvailable(project: Project): Boolean {
...
+
+    private fun parseMavenVersion(output: String?): String? {
+        if (output == null) return null
+        val mavenVersionIndex = output.indexOf("Apache Maven")
+        if (mavenVersionIndex == -1) return null
+        return try {
+            val mavenVersionString = output.slice(IntRange(mavenVersionIndex + 13, output.length - 1))
+            mavenVersionString.slice(IntRange(0, output.indexOf(' ') - 1))
+        } catch (e: StringIndexOutOfBoundsException) {
+            LOG.error(e) { "Failed to parse Maven version" }
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt:348]:
```suggestion
            LOG.error(e) { "Failed to parse Maven version from output: $output" }
```

**Review by rli — APPROVED** [review]

**Review by damntrecky — APPROVED** [review]

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt [hunk 7]
```diff
@@ -285,3 +294,59 @@ fun isCodeModernizerAvailable(project: Project): Boolean {
 }
 
 fun isGradleProject(project: Project) = !GradleSettings.getInstance(project).linkedProjectsSettings.isEmpty()
+
```
Comment (damntrecky) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt:297]: This is fine for now. Maybe something like `TransformLogWrapper` would be more explicit. But I think its clear why these functions are inside if they need access to logger.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt [hunk 8]
```diff
@@ -285,3 +294,59 @@ fun isCodeModernizerAvailable(project: Project): Boolean {
...
+object TransformUtils {
+    private val LOG = getLogger<TransformUtils>()
+
+    fun getJavaVersionFromProjectSetting(project: Project): String? = project.tryGetJdk()?.toString()
```
Comment (damntrecky) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt:301]: Nit. Doesn't need access to logger.

Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt:301]: you can also do `LoggerFactory.getLogger("...")` if it's a util-only file and you need a logger that isn't tied to a class/object

Comment (tincheng) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/Utils.kt:301]: I see, that is much better!

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt [hunk 9]
```diff
@@ -285,3 +296,55 @@ fun isCodeModernizerAvailable(project: Project): Boolean {
 }
 
 fun isGradleProject(project: Project) = !GradleSettings.getInstance(project).linkedProjectsSettings.isEmpty()
+
+fun getJavaVersionFromProjectSetting(project: Project): String? = project.tryGetJdk()?.toString()
+
+fun getMavenVersions(project: Project): String {
```
Comment (dhasani23) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt:302]: Ignore -- just putting this comment here so I can link to it

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt [hunk 10]
```diff
@@ -285,3 +296,55 @@ fun isCodeModernizerAvailable(project: Project): Boolean {
...
+            LOG.error(e) { "Failed to fetch $mavenCommand version" }
+        }
+        return null
+    }
+
+    // Get local maven version
+    val localMavenVersion: String? = getVersion("mvn")
+
+    // Get wrapper maven version
+    val mvnw = if (SystemInfo.isWindows) "./mvnw.cmd" else "./mvnw"
```
Comment (dhasani23) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt:326]: Ignore -- just putting this comment here so I can link to it

**Pull request merged**